### PR TITLE
Let pointer comparison compare absolute addresses

### PR DESCRIPTION
**Pull request by aqjune**
*NOTE: This description is out-of-date; see the comments below

---

This has two changes:

- Let icmp ptr1, ptr2 compare their absolute addresses if either points to the null block.

- Add refinement between ptr and null if ptr.getAddress() == 0. This allows
```
if (ptr == null) use(ptr);
=>
if (ptr == null) use(null);
```

Both are based on the twin memory semantics.

Unfortunately, this patch isn't sufficient to support Transforms/InstSimplify/and-or-icmp-nullptr.ll that was reported by @LebedevRI ... :( Currently icmp ult/ugt... does not consider block sizes, so it still says it's wrong.

BTW, It is amazing that precisely defining pointer comparison is a pretty hard job.

---

### CODE REVIEW

**Comment by LebedevRI** [comment]
Thank you for looking into this!

File: tests/alive-tv/ptrcmp/lifetime-cmp.srctgt.ll [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+	%ip = ptrtoint i32* %p to i64
+	%iq = ptrtoint i32* %q to i64
+	%guard = icmp eq i64 %ip, %iq
+	br i1 %guard, label %COMPARE, label %EXIT
+
+COMPARE:
+	call void @g(i1 false)
+	call void @g(i1 false)
+	br label %EXIT
```
Comment (aqjune) [inline tests/alive-tv/ptrcmp/lifetime-cmp.srctgt.ll:47]: This should fail but succeeds due to #385 .

**Comment by aqjune** [comment]
(As discussed with Nuno) we chose to simply compare the integer addresses of pointers rather than adopting the definition of the twin memory paper.

The reason is as follows:

1. If we make comparison of pointers to different blocks return nondeterministic value (as defined in the twin memory paper), the comparison with null becomes problematic:
 
    - If comparison with null just compares underlying integer addresses, it breaks the and-or-icmp-null optimization. In target, comparison with null is removed, so the result becomes nondeterministic.

    - If comparison with null returns nondeterministic value, it does not explain nonnull attribute; If `icmp ne ptr, null` holds in a branch condition, (int)ptr should not be zero.
 
So the comparison of two different blocks should yield a deterministic value.
 
2. The twin memory's pointer comparison semantics does not support icmp (gep p, i), (gep q, i) to icmp p, q.

**Comment by aqjune** [comment]
Resolved LLVM unit test failures:

```
Transforms/InstCombine/and-or-icmp-nullptr.ll
Transforms/InstCombine/assume.ll
Transforms/InstCombine/cast_ptr.ll
Transforms/InstCombine/icmp.ll
Transforms/InstCombine/or.ll
Transforms/InstSimplify/and-or-icmp-nullptr.ll
```

So finally `and-or-icmp-nullptr.ll` is supported..!

Added failures:
```
Transforms/GVN/condprop.ll : propagation of pointer equality 
Transforms/CorrelatedValuePropagation/phi-common-val.ll : replaces null with p s.t. (int)p == 0
Transforms/InstCombine/gep-custom-dl.ll : icmp (gep p, i), (gep q, i) => icmp p, q is incorrect in some data layout
```

Timeouts: 868 -> 876

To resolve some regressions, other than updating pointer comparison, I had to add two things:

- Refinement from p to null if (int)p is zero. This is allowed in the twin memory paper
- free(p) is no-op if (int)p is zero. This was needed to hoist free(p) out of if(p) branch.

**Comment by aqjune** [comment]
One regression: alive-tv/memory/calloc-overflow.src.ll raises precondition is false error, but I guess this is an error in the encoding of calloc...

**Pull request closed without merging**